File: app/src/main/java/com/playsho/android/utils/DeviceUtils.java
package com.playsho.android.utils;

import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.os.Build;
import android.provider.Settings;
import android.util.Base64;

import androidx.annotation.Nullable;


import com.playsho.android.base.ApplicationLoader;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.UUID;

/**
 * Utility class for retrieving information about the device and application.
 */
public class DeviceUtils {

    /**
     * Number of hashed bytes used in generating app hash.
     */
    public static final int NUM_HASHED_BYTES = 9;

    /**
     * Number of base64 characters used in generating app hash.
     */
    public static final int NUM_BASE64_CHAR = 11;

    /**
     * Hash type for SHA-256 algorithm.
     */
    private static final String HASH_TYPE_SHA_256 = "SHA-256";

    /**
     * Operating System identifier for Android.
     */
    public static final String OS = "android";

    /**
     * Retrieves the model of the device.
     *
     * @return The device model.
     */
    public static String getModel() {
        return Build.MODEL;
    }

    /**
     * Retrieves the manufacturer of the device.
     *
     * @return The device manufacturer.
     */
    public static String getManufacturer() {
        return Build.MANUFACTURER;
    }

    /**
     * Retrieves the brand of the device.
     *
     * @return The device brand.
     */
    public static String getBrand() {
        return Build.BRAND;
    }

    /**
     * Retrieves the Android OS version of the device.
     *
     * @return The Android OS version.
     */
    public static String getOsVersion() {
        return Build.VERSION.RELEASE;
    }

    /**
     * Generates a unique identifier for the device based on its fingerprint and Android ID.
     *
     * @return The device unique identifier.
     */
    public static String getDeviceUniqueId() {
        @SuppressLint("HardwareIds") String androidId = Settings.Secure.getString(
                ApplicationLoader.getAppContext().getContentResolver(), Settings.Secure.ANDROID_ID);
        String fingerprint = Build.FINGERPRINT;
        if (androidId == null || "9774d56d682e549c".equals(androidId)) {
            androidId = UUID.randomUUID().toString();
        }
        return fingerprint.concat("/").concat(androidId);
    }

    /**
     * Retrieves the app signatures using SHA-256 hashing algorithm.
     *
     * @return List of app signatures.
     */
    public static ArrayList<String> getAppSignatures() {
        ArrayList<String> appCodes = new ArrayList<>();
        try {
            String packageName = ApplicationLoader.getAppContext().getPackageName();
            PackageManager packageManager = ApplicationLoader.getAppContext().getPackageManager();
            Signature[] signatures = packageManager.getPackageInfo(packageName, PackageManager.GET_SIGNATURES).signatures;
            for (Signature signature : signatures) {
                String hash = getAppHash(packageName, signature.toCharsString());
                if (hash != null) {
                    appCodes.add(String.format("%s", hash));
                }
            }
        } catch (PackageManager.NameNotFoundException ignore) {
            // Ignore exception if package name is not found.
        }
        return appCodes;
    }

    /**
     * Generates a hash for the app using SHA-256 algorithm.
     *
     * @param packageName The package name of the app.
     * @param signature   The app signature.
     * @return The generated app hash.
     */
    @Nullable
    private static String getAppHash(String packageName, String signature) {
        String appInfo = packageName + " " + signature;
        try {
            MessageDigest messageDigest = MessageDigest.getInstance(HASH_TYPE_SHA_256);
            messageDigest.update(appInfo.getBytes(StandardCharsets.UTF_8));
            byte[] hashSignature = messageDigest.digest();
            hashSignature = Arrays.copyOfRange(hashSignature, 0, NUM_HASHED_BYTES);
            String base64Hash = Base64.encodeToString(hashSignature, Base64.NO_PADDING | Base64.NO_WRAP);
            base64Hash = base64Hash.substring(0, NUM_BASE64_CHAR);
            return base64Hash;
        } catch (NoSuchAlgorithmException ignore) {
            // Ignore exception if SHA-256 algorithm is not available.
        }
        return null;
    }
}

File: app/src/main/java/com/playsho/android/utils/DeviceUtils.kt
package com.playsho.android.utils

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.content.Context
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.os.BatteryManager
import android.os.Build
import android.provider.Settings
import android.telephony.TelephonyManager
import android.util.Log
import androidx.core.content.ContextCompat.getSystemService
import com.playsho.android.base.ApplicationLoader
import com.playsho.android.data.Device
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*

/**
 * Utility class for retrieving information about the device and application.
 */
object DeviceUtils {

    /**
     * Operating System identifier for Android.
     */
    const val OS = "android"

    /**
     * TelephonyManager instance for network-related information
     */
    private val telephonyManager: TelephonyManager? by lazy {
        ApplicationLoader.context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
    }

    /**
     * Retrieves the version name of the application.
     *
     * @return The version name of the application.
     */
    fun getAppVersionName(): String {
        val packageManager = ApplicationLoader.context.packageManager
        val packageName = ApplicationLoader.context.packageName
        return try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            packageInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            "ERR"
        }
    }

    fun getAppVersionCode(): Int {
        val packageManager = ApplicationLoader.context.packageManager
        val packageName = ApplicationLoader.context.packageName
        return try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                packageInfo.longVersionCode.toInt()
            } else {
                packageInfo.versionCode
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            -1 // Return -1 as an error code
        }
    }

    /**
     * Retrieves the first installation time of the application.
     *
     * @return The first installation time of the application.
     */
    fun getFirstInstallTime(): Long {
        val packageManager = ApplicationLoader.context.packageManager
        val packageName = ApplicationLoader.context.packageName
        return try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            packageInfo.firstInstallTime
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            0
        }
    }

    /**
     * Retrieves the last update time of the application.
     *
     * @return The last update time of the application.
     */
    fun getLastUpdateTime(): Long {
        val packageManager = ApplicationLoader.context.packageManager
        val packageName = ApplicationLoader.context.packageName
        return try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            packageInfo.lastUpdateTime
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            0
        }
    }

    /**
     * Retrieves the network operator name.
     *
     * @return The network operator name.
     */
    fun getNetworkOperatorName(): String {
        return telephonyManager?.networkOperatorName ?: ""
    }

    /**
     * Retrieves the percentage of free memory on the device.
     *
     * @return The percentage of free memory.
     */
    fun getFreeMemoryPercentage(): Int {
        val activityManager = ApplicationLoader.context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val memoryInfo = ActivityManager.MemoryInfo()
        activityManager.getMemoryInfo(memoryInfo)
        return ((memoryInfo.availMem.toDouble() / memoryInfo.totalMem) * 100.0).toInt()
    }

    fun getSimCountryIso(): String {
        return telephonyManager?.simCountryIso ?: ""
    }

    fun getSimOperator(): String {
        return telephonyManager?.simOperator ?: ""
    }

    fun getBatteryPercentage(): Int {
        val batteryManager = ApplicationLoader.context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        return batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    }

    fun getNetworkClass(): String {
        val cm = ApplicationLoader.context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val info = cm.activeNetworkInfo
        if (info == null || !info.isConnectedOrConnecting)
            return "NOT_CONNECTED" // not connected
        return when (info.type) {
            ConnectivityManager.TYPE_WIFI -> "WIFI"
            ConnectivityManager.TYPE_MOBILE -> {
                when (info.subtype) {
                    TelephonyManager.NETWORK_TYPE_GPRS,
                    TelephonyManager.NETWORK_TYPE_EDGE,
                    TelephonyManager.NETWORK_TYPE_CDMA,
                    TelephonyManager.NETWORK_TYPE_1xRTT,
                    TelephonyManager.NETWORK_TYPE_IDEN,
                    TelephonyManager.NETWORK_TYPE_GSM -> "2G"
                    TelephonyManager.NETWORK_TYPE_UMTS,
                    TelephonyManager.NETWORK_TYPE_EVDO_0,
                    TelephonyManager.NETWORK_TYPE_EVDO_A,
                    TelephonyManager.NETWORK_TYPE_HSDPA,
                    TelephonyManager.NETWORK_TYPE_HSUPA,
                    TelephonyManager.NETWORK_TYPE_HSPA,
                    TelephonyManager.NETWORK_TYPE_EVDO_B,
                    TelephonyManager.NETWORK_TYPE_EHRPD,
                    TelephonyManager.NETWORK_TYPE_HSPAP,
                    TelephonyManager.NETWORK_TYPE_TD_SCDMA -> "3G"
                    TelephonyManager.NETWORK_TYPE_LTE,
                    TelephonyManager.NETWORK_TYPE_IWLAN,
                    19 -> "4G" // LTE_CA
                    TelephonyManager.NETWORK_TYPE_NR -> "5G"
                    else -> "?"
                }
            }
            else -> "?"
        }
    }

    fun isAirplaneModeOn(): Boolean {
        return Settings.Global.getInt(
            ApplicationLoader.context.contentResolver,
            Settings.Global.AIRPLANE_MODE_ON, 0
        ) != 0
    }

    fun getModel(): String {
        return Build.MODEL
    }

    fun getManufacturer(): String {
        return Build.MANUFACTURER
    }

    fun getBrand(): String {
        return Build.BRAND
    }


    fun getOsVersion(): String {
        return Build.VERSION.RELEASE
    }

    fun createDevice():Device{
        return Device(
            brand = this.getBrand().lowercase(),
            os = "android",
            model =  this.getModel(),
            secret = this.getDeviceUniqueId(),
            manufacturer = this.getManufacturer(),
            osVersion = this.getOsVersion(),
            appVersionName = this.getAppVersionName(),
            appVersion = this.getAppVersionCode(),
            lastUpdateAt = this.getLastUpdateTime(),
            firstInstallAt = this.getFirstInstallTime(),
        )
    }

    @SuppressLint("HardwareIds")
    fun getDeviceUniqueId(): String {
        val androidId = Settings.Secure.getString(
            ApplicationLoader.context.contentResolver,
            Settings.Secure.ANDROID_ID
        ) ?: UUID.randomUUID().toString()
        val fingerprint = Build.FINGERPRINT
        return "$fingerprint/$androidId"
    }

}